### DISCUSSION TITLE
Offline QR Code Verification For Certificate

### DISCUSSION THREAD

**Discussion by priyanka-TL** [Sunbird-Ed/Community#51] [General Discussion]
For the project use case on the Manage Learn side, we are planning to integrate Sunbird RC for certificate generation similar to the way it's done in courses. Refer JIRA - https://project-sunbird.atlassian.net/browse/SB-28231

We have so far succeeded in generating a certificate based on our needs using Sunbird RC local. We have raised a [discussion](https://github.com/orgs/Sunbird-Lern/discussions/21)  in Sunbird Lern to get clarification on how we do it on the platform.

Now we need to understand How does QR code verification happen for courses offline?
From Reshmi we understood that verification happened offline on the client side. If you can help us understand how it happening offline

cc - @aks30 @vijiurs @kiranharidas187 @varsha-sl @Prateek-slokam @aishwaryashikshalokam

**Comment by swayangjit** [comment]
@priyanka-TL On client side it's not completely offline as the public key required for verification needs to be fetched. So first verification will be online and after that, it will be offline. For reference, you can check [this piece of code](https://github.com/Sunbird-Ed/sunbird-client-services/blob/7733490c80ac4962e814310f8d7ad8bc48284e12/src/services/certificate/implementation/certificate-service-impl.ts#L316) for verification.

**Reply by aks30** [reply]
@swayangjit since we don't know frontend code. Can you help with the steps (in English) the app takes to verify the certificate on QR code scan?

**Reply by luckynvdu** [reply]
@aks30 @priyanka-TL The current feature allows user to access the certificate online once then they can access it offline. The flow would be as follows - 

1. User receives the certificate via soft or physical copy, scans the QR code while being online
2. if the QR code is valid then it will show the user a popup with the valid user's certificate attributes
3. If the QR code is invalid then it will show an error popup
4. If the user has accessed the attributes online once (post scanning) then the user can access it offline afterward

@sharathkashyap @swayangjit

**Reply by priyanka-TL** [reply]
@luckynvdu when scanning the QR code which API will call?

**Reply by swayangjit** [reply]
@priyanka-TL This depends on the QR code. If that's a certificate QR  then to verify the certificate we don't need any API. But the parameters(publickey) required for the verification process are fetched from the API.

**Reply by aks30** [reply]
@swayangjit if you can expand on it in the below fashion, it would help us with what we are looking for - 

1. The user scans the QR code (any QR code - certificate, course etc), post which the app derives/extract some information from the code. Exactly what information does the app get, if you can share a couple of samples of the same, I am guessing it would be JSON.
2. After getting the information on Step 1, based on which key in the information of the 1st step does the app know if it's a certificate QR code or course?
3. If the QR code is of type certificate, the app does call an API to do something. Exactly which API or which APIs (in case one after the other there ar multiple based on the whether the scan is happening for the first time or n'th time) is/are called with what req body and what are possible responses from that response and what next steps does the app take based on the same. A visual flow chart if possible would be easy to understand this.
4. Post step 3 maybe if the app is doing anything.

**Reply by swayangjit** [reply]
@aks30  Please post your findings based on our discussion.
